src/Repository/DefaultUnleashProxyRepository.php

```diff
@@ -0,0 +1,207 @@
+use Unleash\Client\Configuration\UnleashConfiguration;
+use Unleash\Client\Configuration\UnleashContext;
+use Unleash\Client\DTO\DefaultProxyFeature;
+use Unleash\Client\DTO\Feature;
+use Unleash\Client\DTO\ProxyFeature;
+
+final class DefaultUnleashProxyRepository implements ProxyRepository
+{
+    public function __construct(
+        private UnleashConfiguration $configuration,
+        private ClientInterface $httpClient,
+        private RequestFactoryInterface $requestFactory
+    ) {
+    }
+
+    public function findFeatureByContext(string $featureName, ?Context $context = null): ?ProxyFeature
+    {
+        $cacheKey = $this->getCacheKey($featureName, $context);
+
+        if ($this->configuration->getCache() !== null && $this->configuration->getCache()->has($cacheKey)) {
+            $cachedFeature = $this->configuration->getCache()->get($cacheKey);
+            if (is_array($cachedFeature)) {
+                $featureData = $this->validateResponse($cachedFeature);
+                if ($featureData !== null) {
+                    return new DefaultProxyFeature($featureData);
+                }
+            }
+        }
+
+        $context ??= new UnleashContext();
+        $featureUrl = $this->configuration->getUrl() . 'frontend/features/' . $featureName;
+        $url = $this->addQuery($featureUrl, $this->contextToQueryString($context));
+
+        $request = $this->requestFactory->createRequest('GET', $url)
+            ->withHeader('Content-Type', 'application/json')
+            ->withHeader('Accept', 'application/json');
+
+        foreach ($this->configuration->getHeaders() as $name => $value) {
+            $request = $request->withHeader($name, $value);
+        }
+
+        $apiKey = $this->configuration->getProxyKey();
+        if ($apiKey === null) {
+            // The only way we can get here is if this is manually constructed without the builder
+            // @codeCoverageIgnoreStart
+            throw new LogicException('No api proxy key was specified');
+            // @codeCoverageIgnoreEnd
+        }
+
+        $request = $request->withHeader('Authorization', $apiKey);
+
+        $response = $this->httpClient->sendRequest($request);
+        $body = json_decode($response->getBody(), true);
+        if ($body === null) {
+            return null;
+        }
+        if (is_array($body)) {
+            $featureData = $this->validateResponse($body);
+
+            if ($featureData !== null) {
+                if ($this->configuration->getCache() !== null) {
+                    $this->configuration->getCache()->set($cacheKey, $featureData);
+                }
+
+                return new DefaultProxyFeature($featureData);
+            }
+        }
+
+        return null;
+    }
+
+    private function getCacheKey(string $featureName, ?Context $context): string
+    {
+        if ($context === null) {
+            return $featureName;
+        }
+        $contextHash = md5($this->contextToQueryString($context));
```

Comment:
Maybe I'm far down into details, but shouldn't we use something like [xxh128](https://php.watch/versions/8.1/xxHash) instead of md5? I know the latter is fast but benchmarks are showing xxh more than 15 times faster: https://php.watch/articles/php-hash-benchmark 

Ofc, we can choose another one if we want to support older PHP versions as xxh is available from 8.1 and if we want to continue to support PHP 7.2 or newer we might have to choose something different like crc32c (which will also reduce from 128b to 32b)

Reply:
I don't _think_ this is a major issue. Hashing a small context object is going to be pretty cheap, and this is supported all the way back to the PHP versions we want to target. I'd rather have a few extra cpu cycles in exchange for wide support

Reply:
What about hash collisions? Md5 is pretty prone to those. I usually use `hash('sha512', $stringToHash)`.

Reply:
> What about hash collisions?
That's a decent argument. MD5 is often good enough but I'll happily take a better hash that's widely supported. I've swapped this out for sha512